src/raftstore/store/peer.rs

```diff
@@ -597,34 +598,42 @@ impl Peer {
         pending_peers
     }
 
-    pub fn check_stale_state(&mut self, d: Duration) -> StaleState {
+    pub fn check_stale_state(&mut self) -> StaleState {
         // Updates the `leader_missing_time` according to the current state.
         if self.leader_id() == raft::INVALID_ID {
             if self.leader_missing_time.is_none() {
                 self.leader_missing_time = Some(Instant::now())
             }
         } else if self.is_initialized() {
-            // A peer is considered as in the leader missing state if it's uninitialized or
-            // if it's initialized but is isolated from its leader.
-            // For an uninitialized peer, even if its leader sends heartbeats to it,
-            // it cannot successfully receive the snapshot from the leader and apply the snapshot.
-            // The raft state machine cannot work in an uninitialized peer to detect
-            // if the leader is working.
             self.leader_missing_time = None
         }
 
         // Checks whether the current peer is stale.
-        let duration = match self.leader_missing_time {
-            Some(t) => t.elapsed(),
-            None => Duration::new(0, 0),
-        };
-        if duration >= d {
-            // Resets the `leader_missing_time` to avoid sending the same tasks to
-            // PD worker continuously during the leader missing timeout.
-            self.leader_missing_time = None;
-            return StaleState::ToValidate;
+        match self.leader_missing_time {
+            Some(t) => {
+                let duration = t.elapsed();
+                if duration >= self.cfg.max_leader_missing_duration.0 {
+                    // Resets the `leader_missing_time` to avoid sending the same tasks to
+                    // PD worker continuously during the leader missing timeout.
+                    self.leader_missing_time = None;
+                    StaleState::ToValidate
+                } else if self.is_initialized()
+                    && duration >= self.cfg.allowed_leader_missing_duration.0
+                {
+                    // A peer is considered as in the leader missing state
```

Comment:
The comment here doesn't explain why only return `LeaderMissing` when it is initialized.